File: mpi/exercises/exercise_8/solution.f90
!> @brief Solution to the 'Ordered hello world' MPI exercise.
PROGRAM main
    USE mpi

    IMPLICIT NONE

    INTEGER :: ierror
    INTEGER :: comm_size
    INTEGER :: my_rank
    INTEGER :: i

    CALL MPI_Init(ierror)

    CALL MPI_Comm_size(MPI_COMM_WORLD, comm_size, ierror)
    CALL MPI_Comm_rank(MPI_COMM_WORLD, my_rank, ierror)

    i = 0

    ! I wait for everybody before me
    DO WHILE (i .LT. my_rank)
        CALL MPI_Barrier(MPI_COMM_WORLD, ierror)
        i = i + 1
    END DO

    ! I print my message
    WRITE(*, '(A,I0,A)') '[MPI Process ', my_rank, '] Hello World!'

    ! I wait for everybody after me
    DO WHILE (i .LT. comm_size)
        CALL MPI_Barrier(MPI_COMM_WORLD, ierror)
        i = i + 1
    END DO

    CALL MPI_Finalize(ierror)
END PROGRAM main

File: mpi/exercises/exercise_8/solution.c
#include <stdio.h>
#include <mpi.h>

/**
 * @brief Solution to the "Ordered hello world" MPI exercise.
 **/
int main(int argc, char* argv[])
{
	MPI_Init(&argc, &argv);

	int comm_size;
	MPI_Comm_size(MPI_COMM_WORLD, &comm_size);

	int my_rank;
	MPI_Comm_rank(MPI_COMM_WORLD, &my_rank);

	int i = 0;

	// I wait for everybody before me
	while(i < my_rank)
	{
		MPI_Barrier(MPI_COMM_WORLD);
		i++;
	}

	// I print my message
	printf("[MPI Process %d] Hello World!\n", my_rank);

	// I wait for everybody after me
	while(i < comm_size)
	{
		MPI_Barrier(MPI_COMM_WORLD);
		i++;
	}

	MPI_Finalize();

	return 0;
}